R/merge_sce_list.R

```diff
@@ -0,0 +1,199 @@
+#'   present in the inputted list
+#' @export
+#'
+#' @import SingleCellExperiment
+merge_sce_list <- function(sce_list = list(),
+                           batch_column = "library_id",
+                           barcode_column = "barcode",
+                           retain_coldata_cols = c("sum",
+                                                   "detected",
+                                                   "total",
+                                                   "subsets_mito_sum",
+                                                   "subsets_mito_detected",
+                                                   "subsets_mito_percent",
+                                                   "miQC_pass",
+                                                   "prob_compromised"),
+                           preserve_rowdata_cols = NULL) {
+
+  # Check `sce_list`----------------------
+  if (is.null(names(sce_list))) {
+    stop("Individual SingleCellExperiment objects in `sce_list` must be named.")
+  }
+  if (length(sce_list) < 2) {
+    warning("There are fewer than two SCE objects in the provided `sce_list` so there is nothing to merge.")
+    # Early return:
+    return(sce_list)
+  }
+
+  # Check `retain_coldata_cols` ----------------
+  if (length(retain_coldata_cols) == 0) {
+    warning("All colData will be removed from the the merged SCE.
+            Please check that `retain_coldata_cols` was correctly specified.")
+  }
+
+  # Subset SCEs to shared features and ensure appropriate naming ------------------
+
+  # First, obtain intersection among all SCE objects
+  shared_features <- sce_list |>
+    purrr::map(rownames) |>
+    purrr::reduce(intersect)
+
+  if (length(shared_features) == 0) {
+    stop("There are no shared features among provided SCE objects.
+         They cannot be merged.")
+  }
+
+  # Second, determine all the column names that are present in any SCE so it can
+  #  be created in any missing SCEs with `NA` values
+  all_colnames <- purrr::map(sce_list, ~names(colData(.))) |>
+    unlist() |>
+    unname() |>
+    unique()
+
+  # Check that the `retain_coldata_cols` are present in at least one SCE, and
+  #  error if the column exists nowhere.
+  if (!(any(retain_coldata_cols %in% all_colnames))) {
+    stop("Error: The provided `retain_coldata_cols` are not present in any SCEs.")
+  }
+
+  # Prepare SCEs
+  sce_list <- purrr::imap(sce_list,
+                          prepare_sce_for_merge,
+                          batch_column,
+                          barcode_column,
+                          shared_features,
+                          # ensure that barcodes are retained
+                          c(barcode_column, retain_coldata_cols),
+                          preserve_rowdata_cols,
+                          all_colnames)
+
+
+  # Create the merged SCE from the processed list ------------------
+  merged_sce <- do.call(cbind, sce_list)
```

Comment:
I prefer to name all purrr arguments that are being passed to the function, just in case the underlying function order changes. Which it shouldn't here, but still...

If we want to retain barcodes by default, I think that should be in the main argument list (commented there as well)

Since I added a lot of verbiage, I added a pipe for formatting...

```suggestion
  # Prepare SCEs
  sce_list <- sce_list |>
    purrr::imap(prepare_sce_for_merge,
                batch_column = batch_column,
                barcode_column =  barcode_column,
                shared_features = shared_features,
                retain_coldata_cols = retain_coldata_cols,
                preserve_rowdata_cols = preserve_rowdata_cols,
                expected_coldata_names = all_colnames) 


  # Create the merged SCE from the processed list ------------------
  merged_sce <- do.call(cbind, sce_list)
```